stream list: Prevent creating streams with the same name

ungps opened the pull request:
Display toast message when user tries to create a stream with the
same as an already existing one.

Fix #3447.

ungps commented:
I am not sure whether this is the best approach. Instead of showing a toast message, we could display an error above the "Create" button. This is the approach followed in the realm screen when we are trying to connect with an invalid organisation URL (like "a.zulipchat.com").

ungps commented:
I force pushed some changes because:

* I found it hard to spot the toast message when the keyboard is up. I think a red error message is more suitable in this case.

* The warning should also appear when we try to edit a stream, not only when we create a stream.

gnprice commented:
Thanks @ungps for sending this!

A few comments:
* I agree that a red error message sounds better than a toast.
* For a PR like this where UI is an important part, please post screenshots in the PR thread; they help a lot for discussion/review.
* I think adding a separate property to the state isn't the preferred React-idiomatic way to do this.
  * One UI consequence is that this version gives an error only when you go hit the submit button -- potentially after typing a bunch of description. Better would be when you blur the name field, or maybe even while you're typing (so in `handleNameChange`).
  * This post looks to have a pretty good description of what I think is a more preferred way to handle this: https://medium.freecodecamp.org/how-to-use-reacts-controlled-inputs-for-instant-form-field-validation-b1c7b033527e
* Does this behave right when editing an existing stream? I think it may not. Be sure to test and think about that case too (which also uses this code, as its name suggests.)

ungps commented:
Thank you @gnprice for review. I pushed a new version. 

 * In both create / edit stream page, the error message is not displayed and the button is disabled when the "Name" field is empty.
 * In edit stream page, the error message is not displayed and the button is disabled when name field wasn't modified.  If description, access (private / public) or the case of the stream name was modified, the button is enabled (the error message still hidden)

Previews:

https://drive.google.com/file/d/13gJJwZvAV-5FLo6PETmw5vruVa-KDzXZ/view?usp=sharing

https://drive.google.com/file/d/16ayxOeDiLQ7Y1H2IUMhUYRmDJvNyuddU/view?usp=sharing

borisyankov commented on src/streams/EditStreamCard.js:
```diff
@@ -47,8 +53,12 @@ export default class EditStreamCard extends PureComponent<Props, State> {
+    const { name, description, isPrivate } = this.state;
+
+    const isInvalidStreamName =
+      caseInsensitiveCompareFunc(name, initialValues.name) !== 0
```
Nice that you are reusing a function but this seems a bit convolutrd.
hat would read better as:
```js
name.toLowerCase() === initialValues.name.toLowerCase()
```

That being said, I am not sure this part is even needed.
On a new stream creation, you are comparing with `''`, and on stream edit, you are comparing with stream's previous name which should be valid.

ungps replied:
Thank you @borisyankov for review.

Indeed, this condition is a little bit counterintuitive, but without it, we would not allow user to change stream name case (ie from "New_Stream" to "NEW_STREAM"). Moreover, the error message would be displayed even if the user did not change stream name field.

borisyankov commented on src/streams/EditStreamCard.js:
```diff
@@ -47,8 +53,12 @@ export default class EditStreamCard extends PureComponent<Props, State> {
+    const { initialValues, isNewStream, streams } = this.props;
+    const { name, description, isPrivate } = this.state;
+
+    const isInvalidStreamName =
```
Often I prefer the positive variant `isValid` as I find it reads better and we don't have to do the double negative to achieve valid `isValid` vs `!isInvalid`.

ungps replied:
I changed it from negative to positive logic.

borisyankov commented on src/streams/EditStreamCard.js:
```diff
@@ -75,10 +86,18 @@ export default class EditStreamCard extends PureComponent<Props, State> {
-          disabled={name.length === 0}
+          disabled={
+            name.length === 0
+            || isMatch(initialValues, { name, description, invite_only: isPrivate })
```
This looks like a lot of logic put in a property value assignment. Would be better as a flag near the `isInvalidStreamName` with a good name like `canUpdateStream` or any other appropriate and helpful name.

borisyankov replied:
`isMatch` looks like overcomplicating code that can be implemented with few `===`s.
While 'lodash' is popular and we do not shy away from using it, I definitely had to go an read the documentation to be able to understand if the code is correct.

This code looks simpler and more exact as:
````js
initialValues.name === name 
&& initialValues. description === description  
&& initialValues. invite_only === isPrivate
```

ungps replied:
Makes sense. I pushed a new version. Thank you!

borisyankov commented on src/streams/EditStreamCard.js:
```diff
@@ -75,10 +90,14 @@ export default class EditStreamCard extends PureComponent<Props, State> {
           style={styles.marginTop}
           text={isNewStream ? 'Create' : 'Update'}
-          disabled={name.length === 0}
+          disabled={canUpdateStream}
```
You meant to do a negative here, right?
`disabled={!canUpdateStream}`

And probably the logic would need changing then?

ungps replied:
Right. Sorry. I force pushed a new version.

gnprice commented:
Thanks again @ungps !

> In edit stream page, the error message is not displayed and the button is disabled when name field wasn't modified. If description, access (private / public) or the case of the stream name was modified, the button is enabled (the error message still hidden)

* This sounds like a basically independent change, right? Let's make this a separate commit, either before or after.

  Then that change can also be explained in its own commit message. Everything we change in a commit should be covered by something described in the commit message -- you could satisfy that by adding a mention of this change to the one commit message, but it's easier to understand if it's separated out.

* Nit: Let's use a name like `canSubmit` rather than `canUpdateStream`. The latter sounds like it might be referring to the contrast between "Update" and "Create", which after all appears in the user text on the very same button.

* Thanks for the videos -- but would you also post an actual screenshot, right here in the thread (so the image shows up right here on the GitHub PR page)? That makes it a lot easier to see and discuss. (In fact, one of the two videos doesn't even load for me.)

  In particular I'd like to see a screenshot of the error message -- that's a new bit of design introduced by this PR, so we'll want it to be part of the discussion.

ungps commented:
Thank you too, @gnprice ! I pushed new changes to

* split the commit into two commits
* Changed `canUpdateStream` to `canSubmit` because is more suitable.
* I am pretty crowded with homework assignments at university and I couldn't keep up with #3450; but I hope I fixed merge issues correctly.    

Below are a couple of screenshots (with caption):

Creating new stream - name field not filled (no error message, button disabled)
![1](https://user-images.githubusercontent.com/24317622/56323124-91fe3480-6173-11e9-9a0f-5ace05915270.png)

Creating new stream - stream name already in-use (display error message, disable button)
![2](https://user-images.githubusercontent.com/24317622/56323125-91fe3480-6173-11e9-9e12-963b4be28d8d.png)

Editing stream - no field changed (no error message, button disabled)
![3](https://user-images.githubusercontent.com/24317622/56323126-9296cb00-6173-11e9-895a-c1a41dda71af.png)

Editing stream - name changed in upper case (no error message, button enabled)
![4](https://user-images.githubusercontent.com/24317622/56323127-9296cb00-6173-11e9-83c7-1da020ad1527.png)

Editing stream - stream name ("announce") already in use (error displayed, button disabled)
In both night mode / day mode:
![5](https://user-images.githubusercontent.com/24317622/56323128-9296cb00-6173-11e9-9834-c7ddf2255d10.png)
![6](https://user-images.githubusercontent.com/24317622/56323129-9296cb00-6173-11e9-9b06-01a18e145d94.png)

gnprice commented:
Thanks @ungps for those screenshots!

One thing they show is that this layout is confusing:
![image](https://user-images.githubusercontent.com/28173/56448170-f7603b80-62c1-11e9-9d88-f2627c6b1b6b.png)

The error is much too far from the field it's about; it's actually closer to the "Description" field below, so it looks like it's about that.

For some pictures of good design for this situation, search for "error" in the Material guidelines:
https://material.io/design/components/text-fields.html#anatomy
Our text fields aren't Material Design in general (although they'd probably be better if they were); but that may help illustrate something to move toward.

ungps commented:
Thanks again, @gnprice!

Indeed, the message was way too far from the field. I hope this is better now. I also changed the colour of underline in case of error.

![Screenshot_20190427-181058](https://user-images.githubusercontent.com/24317622/56851544-03c74280-6919-11e9-8d0f-a05575bec728.png)
![Screenshot_20190427-181119](https://user-images.githubusercontent.com/24317622/56851545-03c74280-6919-11e9-85e8-79b0b6eb402f.png)
![Screenshot_20190427-181331](https://user-images.githubusercontent.com/24317622/56851546-03c74280-6919-11e9-8809-b5dc15116053.png)
![Screenshot_20190427-181630](https://user-images.githubusercontent.com/24317622/56851547-045fd900-6919-11e9-96d3-5b284139e7ce.png)
